sympy/sets/conditionset.py

```
@@ -196,57 +200,39 @@ def free_symbols(self):
-                        # leave the dummy symbol alone -- a second
-                        # subs will be needed to change the dummy
-                        return self.func(sym, cond, base)
-                    else:
-                        return self.func(new, cond.subs(old, new), base)
-                raise ValueError(filldedent('''
-                    A dummy symbol can only be
-                    replaced with a symbol having the same
-                    assumptions or one having a single assumption
-                    having the same commutativity.
-                '''))
-            # don't target cond: it is there to tell how
-            # the base set should be filtered and if new is not in
-            # the base set then this substitution is ignored
-            return self.func(sym, cond, base)
-        else:
-            cond = self.condition.subs(old, new)
-            base = self.base_set.subs(old, new)
-            # The condition may have become true due to assumptions
-            # on 'sym'. In order for .subs() to be consistent with
-            # __new__ we *don't* check if 'sym' actually belongs to
-            # 'base'. In other words: assumptions are ignored.
-            return self.func(self.sym, cond, base)
+        insym = old in flatten([sym])
+        # prioritize changing a symbol in the base
+        newbase = base.subs(old, new)
+        if newbase != base:
+            if not insym:
+                # XXX should we ignore the subs if TypeError is
+                # raised and just return self? e.g. if cond is
+                # x < 1 then subs(x, I) raises TypeError
+                cond = cond.subs(old, new)
+            return self.func(sym, cond, newbase)
+        if insym:
+            if not getattr(new, '_diff_wrt', False):
+                # it's not symbol-like so don't allow this
+                # substitution -- or should we raise an error?
+                return self
```

I guess this could be something like:
```julia
In [1]: ConditionSet((x, (y, z)), x < 3, Reals*Reals**2).subs((y, z), t)                                                          
Out[1]: 
⎧                      ⎛ 2⎞          ⎫
⎨(x, t) | (x, t) ∊ ℝ × ⎝ℝ ⎠ ∧ (x < 3)⎬
⎩                                    ⎭
```

---

dimensionality match between signature and base set should be performed at instantions -- that subs should raise an error.

Also, what do you think about
```
ConditionSet(y, y < 4, S.Integers).subs(y, pi)
```
Should that be just EmptySet?

And should
```
>>> ConditionSet(1,x<2,S.Integers)
ConditionSet(1, x < 2, Integers)
```
just be `FiniteSet(1)`?

---

substitution of bound symbols is no longer supported via `subs`